python_modules/libraries/dagster-components/dagster_components/core/defs_module.py

```diff
@@ -0,0 +1,279 @@
+
+from dagster_components.core.component import (
+    Component,
+    ComponentLoadContext,
+    is_component_loader,
+    load_component_type,
+)
+from dagster_components.core.component_key import ComponentKey
+from dagster_components.utils import load_module_from_path
+
+T = TypeVar("T", bound=BaseModel)
+
+
+class ComponentFileModel(BaseModel):
+    model_config = ConfigDict(extra="forbid")
+
+    type: str
+    attributes: Optional[Mapping[str, Any]] = None
+
+
+#########
+# MODULES
+#########
+
+
+@record
+class DefsModule(ABC):
+    path: Path
+
+    @abstractmethod
+    def build_defs(self) -> Definitions: ...
+
+
+@record
+class SubpackageDefsModule(DefsModule):
+    """A folder containing multiple submodules."""
+
+    submodules: Sequence[DefsModule]
+
+    def build_defs(self) -> Definitions:
+        return Definitions.merge(*(submodule.build_defs() for submodule in self.submodules))
+
+
+@record
+class PythonDefsModule(DefsModule):
+    """A module containing python dagster definitions."""
+
+    module: Any  # ModuleType
+
+    def build_defs(self) -> Definitions:
+        definitions_objects = list(find_objects_in_module_of_types(self.module, Definitions))
+        if len(definitions_objects) == 0:
+            return load_definitions_from_module(self.module)
+        elif len(definitions_objects) == 1:
+            return next(iter(definitions_objects))
+        else:
+            raise DagsterInvalidDefinitionError(
+                f"Found multiple Definitions objects in {self.path}. At most one Definitions object "
+                "may be specified per module."
+            )
+
+
+@record
+class ComponentDefsModule(DefsModule):
+    """A module containing a component definition."""
+
+    context: ComponentLoadContext
+    component: Component
+
+    def build_defs(self) -> Definitions:
+        return self.component.build_defs(self.context)
+
+
+#######
+# DECLS
+#######
+
+
+def _parse_component_yaml(path: Path) -> tuple[SourcePositionTree, ComponentFileModel]:
+    component_file_path = path / "component.yaml"
+    source_tree = parse_yaml_with_source_positions(
+        component_file_path.read_text(), str(component_file_path)
+    )
+    return (
+        source_tree.source_position_tree,
+        _parse_and_populate_model_with_annotated_errors(
+            cls=ComponentFileModel, obj_parse_root=source_tree, obj_key_path_prefix=[]
+        ),
+    )
+
+
+@record
+class DefsModuleDecl(ABC):
+    path: Path
+
+    def get_source_position_tree(self) -> Optional[SourcePositionTree]:
+        return None
+
+    @staticmethod
+    def from_path(path: Path) -> Optional["DefsModuleDecl"]:
+        # this defines the priority of the decl types, we return the first one that matches
+        decltypes = (
+            YamlComponentDecl,
+            PythonComponentDecl,
+            PythonModuleDecl,
+            SubpackageDefsModuleDecl,
+        )
+        decl_filter = filter(None, (cls.from_path(path) for cls in decltypes))
```

Comment:
this is cute but also feels like a strange pattern vs something like `decl_type.exists_in(path)`